src/scenes/context.ts

```diff
@@ -7,52 +7,76 @@ const debug = d('telegraf:scenes:context')
 const noop = () => Promise.resolve()
 const now = () => Math.floor(Date.now() / 1000)
 
+export interface SceneSession {
+  current?: string
+  expires?: number
+  state?: object
+}
+
 // eslint-disable-next-line @typescript-eslint/no-namespace
 namespace SceneContext {
-  export interface Options {
+  export interface Options<D = {}> {
     sessionName: string
     ttl?: number
-    default?: any
+    default?: string
+    defaultSession?: D
+  }
+  export interface Extension<S extends SceneSession, C extends Context> {
+    scene: SceneContext<S, C>
   }
+  export type Extended<S extends SceneSession, C extends Context> = C &
+    Extension<S, C>
 }
 
-class SceneContext<TContext extends Context> {
+class SceneContext<S extends SceneSession, C extends Context> {
+  private readonly options: SceneContext.Options<S> = {
+    sessionName: 'session',
+  }
+
   constructor(
-    private readonly ctx: TContext,
-    private readonly scenes: Map<string, BaseScene<TContext>>,
-    private readonly options: SceneContext.Options
-  ) {}
+    private readonly ctx: C,
+    private readonly scenes: Map<string, BaseScene<C>>,
+    options: Partial<SceneContext.Options<S>>
+  ) {
+    this.options = { sessionName: 'session', ...options }
+  }
 
   get session() {
     const sessionName = this.options.sessionName
-    let session = (this.ctx as any)[sessionName].__scenes ?? {}
-    if (session.expires < now()) {
-      session = {}
+    let session: S = (this.ctx as any)[sessionName].__scenes ?? {}
```

Comment:
3 `this.ctx as any`s in this file have to go

Reply:
There is only three. How are you going to construct an object that has exactly one property which can be parametrized? I don't think TS allows us to do this, but I'm happy if you can prove me wrong

Reply:
```ts
C extends Context & { session?: { __scenes: something } }
```

Reply:
That is wrong. You have to parametrize `session` with the `sessionName` argument of the constructor. Session data is not always stored in the same property on the context object.

Reply:
We would maybe need some magic with string literal types as type variable for the property name, mapped types to dynamically construct the object, `keyof` things to make sure everything matches, … and there is zero benefit to the end user as everything is typed already.

If this was possible, we would be able to infer automatically generated context objects from the user in many cases.

Reply:
I got read of that misfeature to simplify typings.

Reply:
Well have a go

Reply:
I mean, everything. Holy cow. How was I not aware of #1131?

I'll have to look into removing a large part of the type parameters everywhere (not just in this module). Takes some time.

Reply:
Turns out I could not remove the parameters as I wanted to. Either way, I also removed the option to specify `sessionName` for scenes and wizards, hence removing the casts to `any`.